component/discovery/kubernetes_crds/kubernetes_crds.go

```diff
@@ -0,0 +1,336 @@
+		opts:             opts,
+		logger:           logger,
+		config:           cfg,
+		discoveryConfigs: map[string]discovery.Configs{},
+		scrapeConfigs:    map[string]*config.ScrapeConfig{},
+	}
+}
+
+func (c *crdManager) apply() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.discovery.ApplyConfig(c.discoveryConfigs)
+	scs := []*config.ScrapeConfig{}
+	for _, sc := range c.scrapeConfigs {
+		scs = append(scs, sc)
+	}
+	err := c.scraper.ApplyConfig(&config.Config{
+		ScrapeConfigs: scs,
+	})
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error applying scrape configs", "err", err)
+	}
+	level.Debug(c.logger).Log("msg", "scrape config was updated")
+}
+
+func (c *crdManager) clearConfigs(kind string, ns string, name string) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	prefix := fmt.Sprintf("%s/%s/%s", kind, ns, name)
+	for k := range c.discoveryConfigs {
+		if strings.HasPrefix(k, prefix) {
+			delete(c.discoveryConfigs, k)
+			delete(c.scrapeConfigs, k)
+		}
+	}
+}
+
+func (c *crdManager) addPodMonitor(pm *v1.PodMonitor) {
+	for i, ep := range pm.Spec.PodMetricsEndpoints {
+		pmc := c.cg.generatePodMonitorConfig(pm, ep, i)
+		c.mut.Lock()
+		c.discoveryConfigs[pmc.JobName] = pmc.ServiceDiscoveryConfigs
+		c.scrapeConfigs[pmc.JobName] = pmc
+		c.mut.Unlock()
+	}
+	c.apply()
+}
+
+func (c *crdManager) addServiceMonitor(sm *v1.ServiceMonitor) {
+	for i, ep := range sm.Spec.Endpoints {
+		smc := c.cg.generateServiceMonitorConfig(sm, ep, i)
+		c.mut.Lock()
+		c.discoveryConfigs[smc.JobName] = smc.ServiceDiscoveryConfigs
+		c.scrapeConfigs[smc.JobName] = smc
+		c.mut.Unlock()
+	}
+	c.apply()
+}
+
+func (c *crdManager) addProbe(p *v1.Probe) {
+	probe := c.cg.generateProbeConfig(p)
+	c.mut.Lock()
+	c.discoveryConfigs[probe.JobName] = probe.ServiceDiscoveryConfigs
+	c.scrapeConfigs[probe.JobName] = probe
+	c.mut.Unlock()
+	c.apply()
+}
+
+func (c *crdManager) onAddPodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	level.Info(c.logger).Log("msg", "found pod monitor", "name", pm.Name)
+	c.addPodMonitor(pm)
+}
+
+func (c *crdManager) onUpdatePodMonitor(oldObj, newObj interface{}) {
+	pm := oldObj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.addPodMonitor(newObj.(*v1.PodMonitor))
+}
+func (c *crdManager) onDeletePodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.apply()
+}
+
+func (c *crdManager) onAddServiceMonitor(obj interface{}) {
+	sm := obj.(*v1.ServiceMonitor)
+	level.Info(c.logger).Log("msg", "found service monitor", "name", sm.Name)
+	c.addServiceMonitor(sm)
+}
+func (c *crdManager) onUpdateServiceMonitor(oldObj, newObj interface{}) {
+	sm := oldObj.(*v1.PodMonitor)
+	c.clearConfigs("serviceMonitor", sm.Namespace, sm.Name)
+	c.addServiceMonitor(newObj.(*v1.ServiceMonitor))
+}
+func (c *crdManager) onDeleteServiceMonitor(obj interface{}) {
+	sm := obj.(*v1.ServiceMonitor)
+	c.clearConfigs("serviceMonitor", sm.Namespace, sm.Name)
+	c.apply()
+}
```

Comment:
It doesn't feel necessary to have three different add/update/delete functions. You could have one of each with a type switch:

```
func (c *crdManager) onAdd(obj interface{}) {
  switch obj := obj.(type) {
  case *v1.ServiceMonitor:
    // ...

  case *v1.PodMonitor:
    // ... 

  case *v1.Probe:
    // ... 

  default:
    // log an error, presumably 
  }
}
```